### PR TITLE
Stream API Homework Done

### PR DESCRIPTION
**Pull request by Alexsizoff7**

### CODE REVIEW

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 1]
```diff
@@ -2,68 +2,114 @@
...
+import java.util.stream.Collectors;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .sorted((a, b) -> b.getFirstName().compareTo(a.getFirstName()))
+                .map(User::getFirstName)
```
Comment (evgheni-panenco) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]:
```
.map(User::getFirstName)
.sorted(Comparator.reverseOrder())
```

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 2]
```diff
@@ -2,68 +2,114 @@
...
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .sorted((a, b) -> {
+                    if (b.getAge().compareTo(a.getAge()) != 0) {
+                        return b.getAge().compareTo(a.getAge());
+                    } else {
+                        return a.getFirstName().compareTo(b.getFirstName());
+                    }
+                })
```
Comment (evgheni-panenco) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]:
```
.sorted(
            Comparator.comparing(User::getAge).reversed()
                .thenComparing(User::getFirstName))
```

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 3]
```diff
@@ -2,68 +2,114 @@
...
+                        return new SimpleEntry<>(null, e1.getValue());
+                    }
+                }).map(Map.Entry::getKey);
     }
 
     @Override
     public List<User> filterBy(final List<User> users, final Predicate<User>... predicates) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(v -> Arrays.stream(predicates).allMatch(f -> f.test(v)))
```
Comment (evgheni-panenco) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java:84]: Nice approach, nothing needed to be changed, just for example another approach that can be used:
`.filter(Stream.of(predicates).reduce(x -> true, Predicate::and))`
The idea is to combine all the preficates, but your option also works.

**Review by evgheni-panenco — APPROVED** [review]

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 1 above)
Comment (Alexsizoff7) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Done

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 2 above)
Comment (Alexsizoff7) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Done

**Pull request merged**